packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart

```diff
@@ -573,6 +564,16 @@ abstract class WebViewClient {
 
 /// The interface to be used when content can not be handled by the rendering engine for [WebView], and should be downloaded instead.
 abstract class DownloadListener {
+  /// Constructs a [DownloadListener].
+  DownloadListener() {
+    if (!_flutterApisHaveBeenSetup) {
+      DownloadListenerFlutterApi.setup(DownloadListenerFlutterApiImpl());
+      _flutterApisHaveBeenSetup = true;
+    }
+  }
+
+  static bool _flutterApisHaveBeenSetup = false;
```

Comment:
Same.

Would it make sense for all this setup to be done in one central location, perhaps?

Reply:
I wasn't sure where to do it at. Options that I was considering were:

1. In the constructor of `WebView` since it is the central class of the plugin. However, this pattern wouldn't be reproducible for plugins where there wasn't a central class.
2. In the constructor of `AndroidWebView`. However, a user may want to use the `android_webview` API without using the shared platform interface.
3. It could be done in the plugin registration when it's ready, so this would just be temporary. I could add a `TODO` if this should be the case.
4. The current solution. I like this one because it only sets the Flutter API when a user starts to use a WebView. I remember that there is a common complaint that Flutter startup is slow because plugin initialization is done for every plugin at startup. This solution follows the idea of initializing only once it is being used.

Reply:
Is there any real benefit for the setup to be incremental at a per-class-within-the-plugin-level though?

Rather than having the same static and `if` in every single class, you could have a `ensurePigeonApis` helper in a separate file that does that check and then calls *all* of the API setup calls. Then each class could call it (unconditionally) where you currently have all of the logic replicated in each class.

Reply:
That's a good point. At this small of a plugin, it probably has little to no benefit. I created `AndroidWebViewFlutterApis.ensureSetUp`.